File: scripts/render-og-image.py
"""Rasterise public/og-image.svg to public/og-image.png.

Facebook, LinkedIn, WhatsApp and X do not render SVG og:image files — they
show no preview at all — so the shared card must be a raster PNG at the
1200x630 they expect.

Regenerate after editing the SVG:

    python3 scripts/render-og-image.py public/og-image.svg public/og-image.png

Only needs a Chromium binary (set CHROME_PATH to override the default) and
the standard library; the PNG crop is done here rather than pulling in an
image dependency for a file that changes once a year.
"""
import os, struct, subprocess, sys, zlib
from pathlib import Path

W, H = 1200, 630
CHROME = os.environ.get("CHROME_PATH") or "/opt/pw-browsers/chromium-1194/chrome-linux/chrome"
svg_path, out_path = Path(sys.argv[1]), Path(sys.argv[2])

tmp = Path(os.environ.get("TMPDIR", "/tmp"))
html = tmp / "_og.html"
html.write_text(
    '<!doctype html><html><head><meta charset="utf-8"><style>'
    '*{margin:0;padding:0;border:0}'
    f'html,body{{width:{W}px;height:{H}px;overflow:hidden}}'
    f'svg{{display:block;width:{W}px;height:{H}px}}'
    "</style></head><body>" + svg_path.read_text(encoding="utf-8") + "</body></html>",
    encoding="utf-8",
)

raw_png = tmp / "_og_raw.png"
# Probe the viewport shortfall instead of hardcoding it.
probe = tmp / "_probe.html"
probe.write_text(
    f'<!doctype html><html><head><style>*{{margin:0;padding:0}}'
    f'html,body{{background:#fff}}'
    f'#p{{width:{W}px;height:{H}px;background:#ff0000}}</style>'
    f'</head><body><div id="p"></div></body></html>',
    encoding="utf-8",
)


def shoot(src: Path, dst: Path, win_h: int) -> None:
    subprocess.run(
        [CHROME, "--headless", "--no-sandbox", "--disable-gpu", "--hide-scrollbars",
         "--force-device-scale-factor=1", f"--window-size={W},{win_h}",
         f"--screenshot={dst}", f"file://{src}"],
        check=True, capture_output=True,
    )


def read_png(path: Path):
    data = path.read_bytes()
    assert data[:8] == b"\x89PNG\r\n\x1a\n"
    idat, pos = b"", 8
    width = height = depth = color = None
    while pos < len(data):
        length = struct.unpack(">I", data[pos:pos + 4])[0]
        ctype = data[pos + 4:pos + 8]
        body = data[pos + 8:pos + 8 + length]
        if ctype == b"IHDR":
            width, height, depth, color = struct.unpack(">IIBB", body[:10])
        elif ctype == b"IDAT":
            idat += body
        pos += 12 + length
    channels = {0: 1, 2: 3, 4: 2, 6: 4}[color]
    assert depth == 8, "expected 8-bit PNG"
    return width, height, channels, zlib.decompress(idat)


def unfilter(width, height, channels, raw):
    stride = width * channels
    out, prev, pos = bytearray(), bytearray(stride), 0
    for _ in range(height):
        ftype = raw[pos]; pos += 1
        line = bytearray(raw[pos:pos + stride]); pos += stride
        for i in range(stride):
            a = line[i - channels] if i >= channels else 0
            b = prev[i]
            c = prev[i - channels] if i >= channels else 0
            if ftype == 1: line[i] = (line[i] + a) & 0xFF
            elif ftype == 2: line[i] = (line[i] + b) & 0xFF
            elif ftype == 3: line[i] = (line[i] + (a + b) // 2) & 0xFF
            elif ftype == 4:
                p = a + b - c
                pa, pb, pc = abs(p - a), abs(p - b), abs(p - c)
                pr = a if (pa <= pb and pa <= pc) else (b if pb <= pc else c)
                line[i] = (line[i] + pr) & 0xFF
        out += line
        prev = line
    return bytes(out)


def write_png(path: Path, width, height, channels, pixels):
    stride = width * channels
    raw = b"".join(b"\x00" + pixels[y * stride:(y + 1) * stride] for y in range(height))
    color = {1: 0, 3: 2, 4: 6}[channels]
    def chunk(tag, body):
        return struct.pack(">I", len(body)) + tag + body + struct.pack(">I", zlib.crc32(tag + body) & 0xFFFFFFFF)
    path.write_bytes(
        b"\x89PNG\r\n\x1a\n"
        + chunk(b"IHDR", struct.pack(">IIBBBBB", width, height, 8, color, 0, 0, 0))
        + chunk(b"IDAT", zlib.compress(raw, 9))
        + chunk(b"IEND", b"")
    )


# 1. Measure how many rows the viewport actually paints.
probe_png = tmp / "_probe.png"
shoot(probe, probe_png, H)
pw, ph, pch, praw = read_png(probe_png)
ppx = unfilter(pw, ph, pch, praw)
painted = 0
for y in range(ph):
    if ppx[y * pw * pch] > 200 and ppx[y * pw * pch + 1] < 60:
        painted = y + 1
shortfall = H - painted
print(f"viewport paints {painted}/{H} rows -> shortfall {shortfall}px")

# 2. Render tall enough that the real content area is exactly H, then crop.
# A generous margin costs nothing: we crop back to exactly H rows below.
shoot(html, raw_png, H + shortfall + 40)
rw, rh, rch, rraw = read_png(raw_png)
px = unfilter(rw, rh, rch, rraw)
stride = rw * rch
cropped = px[:H * stride]
write_png(out_path, W, H, rch, cropped)
print(f"wrote {out_path} ({out_path.stat().st_size} bytes) from a {rw}x{rh} render")

File: scripts/render-og-image.mjs
#!/usr/bin/env node
//
// Renders public/og-image.svg to public/og-image.png at exactly 1200x630.
//
// Why a PNG at all: Facebook, LinkedIn, WhatsApp and X all ignore an SVG
// og:image. The site shipped only an SVG, so every share of the link showed no
// preview — on a product whose entire promise is social media. Edit the SVG,
// run this, commit both.
//
// Uses the Chromium that Playwright installs (PLAYWRIGHT_BROWSERS_PATH), with
// no extra dependency. Chromium's --window-size includes browser chrome, so we
// render taller than needed and crop the top 630 rows back off.
import { execFileSync } from "node:child_process";
import { existsSync, mkdtempSync, readFileSync, writeFileSync, readdirSync } from "node:fs";
import { tmpdir } from "node:os";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { deflateSync, inflateSync } from "node:zlib";

const ROOT = join(dirname(fileURLToPath(import.meta.url)), "..");
const WIDTH = 1200;
const HEIGHT = 630;
// Extra height to absorb the non-content part of the Chromium window.
const CHROME_OVERHEAD = 120;

function findChromium() {
  const base = process.env.PLAYWRIGHT_BROWSERS_PATH || "/opt/pw-browsers";
  if (!existsSync(base)) return null;
  for (const entry of readdirSync(base)) {
    if (!entry.startsWith("chromium-")) continue;
    const candidate = join(base, entry, "chrome-linux", "chrome");
    if (existsSync(candidate)) return candidate;
  }
  return null;
}

// --- Minimal PNG crop (8-bit RGB/RGBA, non-interlaced) ---------------------

function readPng(path) {
  const data = readFileSync(path);
  let pos = 8;
  let header = null;
  const idat = [];
  while (pos < data.length) {
    const length = data.readUInt32BE(pos);
    const type = data.subarray(pos + 4, pos + 8).toString("ascii");
    const body = data.subarray(pos + 8, pos + 8 + length);
    if (type === "IHDR") {
      header = { width: body.readUInt32BE(0), height: body.readUInt32BE(4), depth: body[8], colorType: body[9] };
    } else if (type === "IDAT") {
      idat.push(body);
    }
    pos += 12 + length;
  }
  if (!header) throw new Error("PNG has no IHDR");
  if (header.depth !== 8) throw new Error(`Unsupported bit depth ${header.depth}`);
  return { header, raw: inflateSync(Buffer.concat(idat)) };
}

function unfilter(raw, width, height, bpp) {
  const stride = width * bpp;
  const out = Buffer.alloc(stride * height);
  let prev = Buffer.alloc(stride);
  let offset = 0;
  for (let y = 0; y < height; y++) {
    const filter = raw[offset++];
    const line = Buffer.from(raw.subarray(offset, offset + stride));
    offset += stride;
    for (let x = 0; x < stride; x++) {
      const a = x >= bpp ? line[x - bpp] : 0;
      const b = prev[x];
      const c = x >= bpp ? prev[x - bpp] : 0;
      if (filter === 1) line[x] = (line[x] + a) & 255;
      else if (filter === 2) line[x] = (line[x] + b) & 255;
      else if (filter === 3) line[x] = (line[x] + ((a + b) >> 1)) & 255;
      else if (filter === 4) {
        const p = a + b - c;
        const pa = Math.abs(p - a);
        const pb = Math.abs(p - b);
        const pc = Math.abs(p - c);
        line[x] = (line[x] + (pa <= pb && pa <= pc ? a : pb <= pc ? b : c)) & 255;
      }
    }
    line.copy(out, y * stride);
    prev = line;
  }
  return out;
}

function chunk(type, body) {
  const out = Buffer.alloc(body.length + 12);
  out.writeUInt32BE(body.length, 0);
  out.write(type, 4, "ascii");
  body.copy(out, 8);
  const crcInput = Buffer.concat([Buffer.from(type, "ascii"), body]);
  out.writeInt32BE(crc32(crcInput), body.length + 8);
  return out;
}

let crcTable = null;
function crc32(buf) {
  if (!crcTable) {
    crcTable = new Int32Array(256);
    for (let n = 0; n < 256; n++) {
      let c = n;
      for (let k = 0; k < 8; k++) c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
      crcTable[n] = c;
    }
  }
  let c = -1;
  for (const byte of buf) c = crcTable[(c ^ byte) & 0xff] ^ (c >>> 8);
  return c ^ -1;
}

function writePng(path, width, height, colorType, bpp, pixels) {
  const stride = width * bpp;
  const raw = Buffer.alloc((stride + 1) * height);
  for (let y = 0; y < height; y++) {
    raw[y * (stride + 1)] = 0; // filter: none
    pixels.copy(raw, y * (stride + 1) + 1, y * stride, (y + 1) * stride);
  }
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(width, 0);
  ihdr.writeUInt32BE(height, 4);
  ihdr[8] = 8;
  ihdr[9] = colorType;
  writeFileSync(
    path,
    Buffer.concat([
      Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]),
      chunk("IHDR", ihdr),
      chunk("IDAT", deflateSync(raw, { level: 9 })),
      chunk("IEND", Buffer.alloc(0)),
    ]),
  );
}

// --- Main -------------------------------------------------------------------

const chromium = findChromium();
if (!chromium) {
  console.error(
    "Chromium not found. Set PLAYWRIGHT_BROWSERS_PATH, or regenerate the PNG\n" +
      "elsewhere and commit it — public/og-image.png must stay 1200x630.",
  );
  process.exit(1);
}

const work = mkdtempSync(join(tmpdir(), "og-"));
const svg = readFileSync(join(ROOT, "public/og-image.svg"), "utf8");
writeFileSync(
  join(work, "og.svg"),
  svg.replace("<svg ", `<svg width="${WIDTH}" height="${HEIGHT}" `),
);
writeFileSync(
  join(work, "og.html"),
  `<!doctype html><html><head><meta charset="utf-8"><style>*{margin:0;padding:0}` +
    `html,body{background:#0F172A}img{display:block;width:${WIDTH}px;height:${HEIGHT}px}` +
    `</style></head><body><img src="og.svg"></body></html>`,
);

const shot = join(work, "shot.png");
execFileSync(
  chromium,
  [
    "--headless",
    "--no-sandbox",
    "--disable-gpu",
    "--hide-scrollbars",
    "--force-device-scale-factor=1",
    `--window-size=${WIDTH},${HEIGHT + CHROME_OVERHEAD}`,
    `--screenshot=${shot}`,
    `file://${join(work, "og.html")}`,
  ],
  { stdio: "ignore" },
);

const { header, raw } = readPng(shot);
const bpp = header.colorType === 6 ? 4 : header.colorType === 2 ? 3 : null;
if (!bpp) throw new Error(`Unsupported colour type ${header.colorType}`);
if (header.width !== WIDTH || header.height < HEIGHT) {
  throw new Error(`Unexpected screenshot size ${header.width}x${header.height}`);
}
const pixels = unfilter(raw, header.width, header.height, bpp);
const out = join(ROOT, "public/og-image.png");
writePng(out, WIDTH, HEIGHT, header.colorType, bpp, pixels.subarray(0, WIDTH * bpp * HEIGHT));
console.log(`Wrote public/og-image.png (${WIDTH}x${HEIGHT}).`);
